app/frontend/src/features/auth/useAuthStore.ts

```diff
@@ -0,0 +1,169 @@
+  const [error, setError] = useState<string | null>(null);
+
+  const authActions = {
+    clearError() {
+      setError(null);
+    },
+    authError(message: string) {
+      setError(message);
+    },
+    async passwordLogin({
+      username,
+      password,
+    }: {
+      username: string;
+      password: string;
+    }) {
+      setError(null);
+      setLoading(true);
+      try {
+        const auth = await service.user.passwordLogin(username, password);
+        setJailed(auth.jailed);
+
+        if (!auth.jailed) {
+          const user = await service.user.getUser(username);
+          setUser(user);
+        }
+        setAuthenticated(true);
+      } catch (e) {
+        setError(e.message);
+      }
+      setLoading(false);
+    },
+    async tokenLogin(loginToken: string) {
+      setError(null);
+      setLoading(true);
+      try {
+        const auth = await service.user.tokenLogin(loginToken);
+        setJailed(auth.jailed);
+
+        if (!auth.jailed) {
+          const user = await service.user.getCurrentUser();
+          setUser(user);
+        }
+
+        setAuthenticated(true);
+      } catch (e) {
+        setError(e.message);
+      }
+      setLoading(false);
+    },
+    async signup(signupArguments: SignupArguments) {
+      setError(null);
+      setLoading(true);
+      try {
+        const auth = await service.user.completeSignup(signupArguments);
+        setJailed(auth.jailed);
+
+        if (!auth.jailed) {
+          const user = await service.user.getCurrentUser();
+          setUser(user);
+        }
+        setAuthenticated(true);
+      } catch (e) {
+        setError(e.message);
+      }
+
+      setLoading(false);
+    },
+    async updateJailStatus() {
+      setError(null);
+      setLoading(true);
+      if (!authenticated) {
+        throw Error("User is not connected.");
+      }
+      try {
+        const isJailed = (await service.jail.getIsJailed()).isJailed;
+        setJailed(isJailed);
+      } catch (e) {
+        setError(e.message);
+      }
+      setLoading(false);
+    },
+    async logout() {
+      setError(null);
+      setLoading(true);
+      try {
+        service.user.logout();
+        setUser(null);
+        setAuthenticated(false);
+      } catch (e) {
+        setError(e.message);
+      }
+      setLoading(false);
+    },
+    async updateUserProfile(userData: UpdateUserProfileData) {
+      const username = user?.username;
+
+      if (!username) {
+        throw Error("User is not connected.");
+      }
+
+      await service.user.updateProfile(userData);
+
+      setUser(await service.user.getUser(username));
+    },
+    async updateHostingPreferences(preferences: HostingPreferenceData) {
+      const username = user?.username;
+
+      if (!username) {
+        throw Error("User is not connected.");
+      }
+
+      await service.user.updateHostingPreference(preferences);
+
+      setUser(await service.user.getUser(username));
+    },
```

Comment:
Probably best to split this up into its own action object as it's not strictly auth related? Then you should be able to memoise the rest of the object whilst avoiding the problem of these two actions reading stale values of `user`.

Reply:
Yes I will do that. As long as we won't need the state in any other actions that come up...

In the next PR I am going to refactor userCache to hooks and take user out of Auth completely, and just leave user ID.